### PR TITLE
Give copy an optional device argument.

### PR DESCRIPTION
**Pull request by gchanan**
This is to avoid having to put the DeviceGuard in core (also it's a nicer API).

### CODE REVIEW

**Comment by gchanan** [comment]
Stacked on top of: https://github.com/pytorch/pytorch/pull/11441
To review: https://github.com/pytorch/pytorch/pull/11467/files/6fc5bb11523cb5d73b7d2cf279c01ea3a2107fc8..HEAD

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/templates/TypeDefault.cpp [hunk 1]
```diff
@@ -22,8 +22,11 @@ Tensor & TypeDefault::copy_(Tensor & self, const Tensor & src, bool non_blocking
   return s_copy_(self, b_src, non_blocking);
 }
 
-Tensor TypeDefault::copy(const Tensor & src, bool non_blocking) const {
-  // TODO(psag): have a DeviceGuard here
+Tensor TypeDefault::copy(const Tensor & src, bool non_blocking, optional<Device> to_device) const {
+  DeviceGuard device_guard;
+  if (to_device.has_value()) {
+    device_guard.set_index(to_device.value().index());
```
[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/templates/TypeDefault.cpp]: define "wrong thing".  I guess the behavior is strange if:
1) you have a CUDA Type, but pass in a non-cuda device
2) you have a non-CUDA Type, but pass in a cuda device.

[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/templates/TypeDefault.cpp]: What do you think about asserting that if the device is present, it matches the type?

[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/templates/TypeDefault.cpp]: If you pass CPU device that somehow has index set to 1, you will then twiddle the current *CUDA* device to 1, even though there's no reason these should be related at all.

Basically, there is something very smelly with the `DeviceGuard` interface. I don't think you should necessarily fix it this PR, but something has to be worked out...

[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/templates/TypeDefault.cpp]: Well, you can't have a CPU device that has index set to 1 (https://github.com/pytorch/pytorch/blob/f2f43ad2dab024ef4010b63ec2faa64e9affcdac/aten/src/ATen/core/Device.h#L35-L38), but what you said does apply to index 0.

**Comment by gchanan** [comment]
closing in favor of https://github.com/pytorch/pytorch/pull/11497.

**Pull request closed without merging**